### PR TITLE
Add virtual columns to timeseries, topN, and groupBy.

### PR DESCRIPTION
**Pull request by gianm**
They were added to "select" in #2511, this adds them to timeseries, topN, and groupBy as well. They remain undocumented for now. See https://groups.google.com/d/topic/druid-development/_Sd78s7yU5U/discussion for more about virtual columns.

### CODE REVIEW

**Comment by jon-wei** [comment]
👍

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java [hunk 1]
```diff
@@ -176,15 +190,16 @@ public TimeseriesQuery withDimFilter(DimFilter dimFilter)
...
-        ", descending=" + isDescending() +
-        ", dimFilter=" + dimFilter +
-        ", granularity='" + granularity + '\'' +
-        ", aggregatorSpecs=" + aggregatorSpecs +
-        ", postAggregatorSpecs=" + postAggregatorSpecs +
-        ", context=" + getContext() +
-        '}';
+           "dataSource='" + getDataSource() + '\'' +
+           ", querySegmentSpec=" + getQuerySegmentSpec() +
+           ", virtualColumns=" + virtualColumns +
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: virtualColumns and descending go in different order elsewhere in this class

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java [hunk 2]
```diff
@@ -211,17 +226,19 @@ public boolean equals(Object o)
     if (granularity != null ? !granularity.equals(that.granularity) : that.granularity != null) {
       return false;
     }
-    if (postAggregatorSpecs != null ? !postAggregatorSpecs.equals(that.postAggregatorSpecs) : that.postAggregatorSpecs != null) {
+    if (aggregatorSpecs != null ? !aggregatorSpecs.equals(that.aggregatorSpecs) : that.aggregatorSpecs != null) {
       return false;
     }
-
-    return true;
+    return postAggregatorSpecs != null
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: IntelliJ can automatically generate `Objects.equals()` (in code generation wizard menu, there is such option in dropdown menu)

File: processing/src/main/java/io/druid/query/topn/TopNQuery.java [hunk 3]
```diff
@@ -67,6 +70,7 @@ public TopNQuery(
   )
   {
     super(dataSource, querySegmentSpec, false, context);
+    this.virtualColumns = virtualColumns == null ? VirtualColumns.EMPTY : virtualColumns;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/TopNQuery.java]: Consider adding helper `VirualColumns.nullToEmpty()`

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: Changed

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java (hunk 1 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: Swapped those here in toString

File: processing/src/main/java/io/druid/query/topn/TopNQuery.java (hunk 3 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQuery.java]: Added that and updated similar call sites

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: Also changed topN and groupBy

**Comment by fjy** [comment]
👍

**Comment by gianm** [comment]
Resolving conflicts with master, will re push soon

**Comment by gianm** [comment]
Pushed with review comments addressed and conflicts fixed, thx @fjy @jon-wei @leventov for review.

**Pull request merged**